pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,275 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/jackc/pgtype"
+
+	"github.com/prometheus/client_golang/prometheus"
+	io_prometheus_client "github.com/prometheus/client_model/go"
+
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/pgmodel/model/pgutf8str"
+	"github.com/timescale/promscale/pkg/pgxconn"
+)
+
+// ExecPlatform To fill this variable in build time, use linker flags.
+// Example: go build -ldflags="-X github.com/timescale/promscale/pkg/telemetry.ExecPlatform=<any_string>" ./cmd/promscale/
+var ExecPlatform string
+
+type engine struct {
+	uuid [16]byte
+	conn pgxconn.PgxConn
+
+	stopTelemetryRoutine context.CancelFunc
+	wgTelemetryRoutine   *sync.WaitGroup
+
+	metricMux sync.RWMutex
+	metrics   []telemetry
+}
+
+type Telemetry interface {
+	StartRoutineAsync()
+	// RegisterMetric registers a Prometheus Gauge or Counter metric for telemetry visitor.
+	// It must be called after creating the telemetry engine.
+	RegisterMetric(columnName string, gaugeOrCounterMetric prometheus.Metric) error
```

Comment:
I wonder if we really need this generic interface for registering telemetry metrics? It seems we already know which metrics are we going to track in our code. I guess that would eliminate some code like `metricsVisitor` where we match on strings and instead we can directly map `metrics` to `telemetryStats`.

Reply:
MetricsVisitor will still be there, as we need to travel though the metrics that we are tracking and get the latest stats, but what can be improved here will be instead of `metrics` being a `[]telemetry`, we can make it a map of `name: prometheus_metric`.

Regarding the `RegisterMetric`, we can register two ways:
1. (the current approach). We know the exact metrics to be monitored. Register those metrics in the telemetry instance passed down the constructor function. This avoids the private metric vars from making them public just for telemetry use-case.
2. In the telemetry package itself, make a var `metrics` and simply import the metrics that you want to monitor from respective packages into this var. However, private metric vars need to be made public.

I feel the first approach is minimalist plus, making a function do this gives the option to error out in runtime, in case a bad metric is asked to be monitored. Second approach does not give the erroring out advantage.